in_toto/verifylib.py

```diff
@@ -239,16 +239,17 @@ def verify_all_steps_signatures(layout, chain_link_dict):
 
   """
   for step in layout.steps:
-    keys_dict = {}
-
     # Find the according link for this step
     key_link_dict = chain_link_dict[step.name]
 
-    # Create the dictionary of keys for this step
-    for pubkeyid in step.pubkeys:
-      keys_dict[pubkeyid] = layout.keys[pubkeyid]
-
     for keyid, link in six.iteritems(key_link_dict):
+      keys_dict = {}
+
+      # Create the dictionary of keys for this step
+      for pubkeyid in step.pubkeys:
+          if pubkeyid == keyid:
+            keys_dict[pubkeyid] = layout.keys[pubkeyid]
```

Comment:
The variable value `keyid` holds the (whole) keyid, determined by the filename infix. So instead of taking all keys authorized for a certain step (`step.pubkeys`) and verifying that the link file is signed by one ore more of those keys, we take the key, identified by `keyid`, check if it is authorized, and then verify if the link was signed with that key.